libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationService.java

```diff
@@ -244,51 +252,63 @@ public void onMilestoneTrigger(List<Milestone> triggeredMilestones, RouteProgres
   public void onUserOffRoute(Location location, boolean userOffRoute) {
     if (userOffRoute) {
       if (location.getTime() > timeIntervalSinceLastOffRoute
         + TimeUnit.SECONDS.toMillis(mapboxNavigation.options().secondsBeforeReroute())) {
-        recentDistancesFromManeuverInMeters.clear();
-        mapboxNavigation.getEventDispatcher().onUserOffRoute(location);
         timeIntervalSinceLastOffRoute = location.getTime();
+        if (mapboxNavigation.getSessionState().lastReroutePosition() == null) {
+          rerouteSessionsStateUpdate();
+        } else {
+          if (TurfMeasurement.distance(mapboxNavigation.getSessionState().lastReroutePosition(),
+            Position.fromLngLat(location.getLongitude(), location.getLatitude()),
+            TurfConstants.UNIT_METERS)
+            > mapboxNavigation.options().minimumDistanceBeforeRerouting()) {
+            mapboxNavigation.getEventDispatcher().onUserOffRoute(location);
+            rerouteSessionsStateUpdate();
+          }
+        }
       }
     } else {
       timeIntervalSinceLastOffRoute = location.getTime();
     }
   }
 
+  private void rerouteSessionsStateUpdate() {
+    recentDistancesFromManeuverInMeters.clear();
+    mapboxNavigation.getEventDispatcher().onUserOffRoute(rawLocation);
+    mapboxNavigation.setSessionState(
+      mapboxNavigation.getSessionState().toBuilder().lastReroutePosition(
+        Position.fromLngLat(rawLocation.getLongitude(), rawLocation.getLatitude())).build()
+    );
+  }
+
   public void rerouteOccurred() {
     mapboxNavigation.setSessionState(mapboxNavigation.getSessionState().toBuilder()
+      .routeProgressBeforeReroute(routeProgress)
       .beforeRerouteLocations(Arrays.asList(
         locationBuffer.toArray(new Location[locationBuffer.size()])))
-      .routeProgressBeforeReroute(routeProgress)
+      .lastRerouteDate(new Date())
       .build());
-    locationBuffer.clear();
-    queuedRerouteEvent = true;
-
-    handler = new Handler();
-    runnable = new Runnable() {
-      @Override
-      public void run() {
-        mapboxNavigation.setSessionState(mapboxNavigation.getSessionState().toBuilder()
-          .afterRerouteLocations(Arrays.asList(
-            locationBuffer.toArray(new Location[locationBuffer.size()])))
-          .build());
-
-        locationBuffer.clear();
-
-        NavigationMetricsWrapper.rerouteEvent(mapboxNavigation.getSessionState(), routeProgress,
-          location);
-        queuedRerouteEvent = false;
-      }
-    };
-    handler.postDelayed(runnable, TWENTY_SECOND_INTERVAL);
+
+    queuedRerouteEvents.add(mapboxNavigation.getSessionState());
+  }
+
+  void sendRerouteEvent(SessionState sessionState) {
+    mapboxNavigation.setSessionState(sessionState.toBuilder()
+      .afterRerouteLocations(Arrays.asList(
+        locationBuffer.toArray(new Location[locationBuffer.size()])))
+      .build());
+
+    NavigationMetricsWrapper.rerouteEvent(sessionState, routeProgress, rawLocation);
```

Comment:
Because sendRerouteEvent is called 20s after the original reroute time, we should be sending the original reroute location not `rawLocation` here. Also, we should override telemetry with the original reroute timestamp rather than have it use the current timestamp.

Reply:
this has been resolved in https://github.com/mapbox/mapbox-navigation-android/pull/386/files#diff-7e57c49bef82db11f00cfa982b5e2e07R313